New syntax

feiranwang opened the pull request.

feiranwang committed:
New syntax for rule annotation, first step

feiranwang committed:
Merge feiran-extension

feiranwang committed:
General rule annotations

feiranwang committed:
New function call syntax

feiranwang committed:
Merge feiran-extension

feiranwang committed:
Refactor conjunctive query and function call rule

feiranwang committed:
Revert to old annotation parsing

feiranwang committed:
MLN style inference rule head

feiranwang committed:
Merge master

netj commented on src/main/scala/org/deepdive/ddlog/DeepDiveLogCompiler.scala:
```diff
@@ -130,18 +129,19 @@ class CompilationState( statements : DeepDiveLog.Program, config : DeepDiveLog.C
+  //     case s: ExtractionRule    => s"extraction_rule_${statements indexOf s}"
+  //     case s: InferenceRule     => s"extraction_rule_${statements indexOf s}"
+  //   }
+  // }
```
Could you remove these commented lines?

Ideally, we should use mnemonic names for each compiled extractor, so user's handwritten extractors can depend on them using it.  Is it easy to generate a unique but stable identifier for each?  This might be easy if we're already grouping by head and producing a union anyway.

feiranwang replied:
Sorry for the commented lines. I'll look into the extractor naming.

netj commented on src/main/scala/org/deepdive/ddlog/DeepDiveLogCompiler.scala:
```diff
@@ -113,10 +112,10 @@ class CompilationState( statements : DeepDiveLog.Program, config : DeepDiveLog.C
-      case InferenceRule(_,_,_,_) => ()
       case fdecl : FunctionDeclaration => function_schema += {fdecl.functionName -> fdecl}
       case f: FunctionCallRule => functionCallList += f
+      case f: InferenceRule => inferenceRules += f
```
Would it be easy to get rid of these `ListBuffer`s using `collect { case f: Type => f }`?

feiranwang replied:
Sure.

netj commented on src/main/scala/org/deepdive/ddlog/DeepDiveLogCompiler.scala:
```diff
@@ -200,19 +198,17 @@ class CompilationState( statements : DeepDiveLog.Program, config : DeepDiveLog.C
       case f : FunctionCallRule => dependencies += {
-        f -> (( Some(f.input) flatMap (stmtByHeadName get _)).toSet.flatten.flatten) }
+        f -> ((( f.q.bodies.flatten collect { case x: BodyAtom => x.name })
+          flatMap (stmtByHeadName get _)).toSet.flatten.flatten) }
```
This seems like a pretty overloaded code to be inline and repeated.  Can we extract it as a function, say `collectUsedRelations(bodies: List[List[Body]]): Set[Statement]`?

feiranwang replied:
Sure.

netj commented on src/main/scala/org/deepdive/ddlog/DeepDiveLogCompiler.scala:
```diff
@@ -446,6 +440,12 @@ class QueryCompiler(cq : ConjunctiveQuery, ss: CompilationState) {
 
+  def generateSQL(useAlias: Boolean) = {
+    val head = generateSQLHead(useAlias)
+    val body = cq.bodies map generateSQLBody mkString("\nUNION ALL\n")
```
Does joining only the bodies with `UNION ALL` correct?  I think each part should start with `SELECT`?  Do we have test cases with more than one body?

feiranwang replied:
I'll look into this one. Looks like a bug here. We have test cases for disjunctive bodies.

netj commented on src/main/scala/org/deepdive/ddlog/DeepDiveLogCompiler.scala:
```diff
@@ -458,7 +458,7 @@ object DeepDiveLogCompiler extends DeepDiveLogHandler {
     var schemas = new ListBuffer[String]()
     for (stmt <- stmts) {
-      if ((stmt.a.name startsWith "dd_new_") && (ss.inferenceRuleGroupByHead contains stmt.a.name)) {
+      if ((stmt.a.name startsWith "dd_new_") && (ss.isQueryTerm(stmt.a.name))) {
```
Looks like here's another idiom that's worth extracting a function for: `isForIncremental(stmt: Statement): Boolean` (or a with clearer name).  I see it's repeated a lot of times below.

netj commented on src/main/scala/org/deepdive/ddlog/DeepDiveLogCompiler.scala:
```diff
@@ -515,7 +515,9 @@ object DeepDiveLogCompiler extends DeepDiveLogHandler {
-        } else if ((ss.schemaDeclarationGroupByHead contains stmt.q.head.name) && (ss.schemaDeclarationGroupByHead(stmt.q.head.name)(0).isQuery) && (stmt.q.head.name startsWith "dd_new_")) {
+        } else if ((ss.schemaDeclarationGroupByHead contains stmt.headName) &&
+          (ss.schemaDeclarationGroupByHead(stmt.headName)(0).isQuery) &&
+          (stmt.headName startsWith "dd_new_")) {
```
Sorry, I now feel like asking too much clean up, for code you just introduced line breaks, but it's pretty painful to decipher what each condition means.  Would have been much nicer if they were like:

``` scala
if (ss.hasSchemaDeclared(stmt) &&
 ss.isQueryTerm(stmt.headName) &&
 ss.isForIncremental(stmt)) {
```

feiranwang replied:
These suggestions looks good.

netj commented on src/main/scala/org/deepdive/ddlog/DeepDiveLogCompiler.scala:
```diff
@@ -603,11 +602,13 @@ object DeepDiveLogCompiler extends DeepDiveLogHandler {
+            case x: Expr    => ExprPattern(x)
+          })
+        }
+        val fakeBody        = headAsBody ++ cqBody
```
This looks like some clever hack, which may be okay for the short term.  However, a comment explaining why we turn a head into a body would be extremely helpful for someone who wants to do the right thing later.

feiranwang replied:
Sure.

netj commented on src/main/scala/org/deepdive/ddlog/DeepDiveLogParser.scala:
```diff
@@ -265,43 +267,82 @@ class DeepDiveLogParser extends JavaTokenParsers {
+  }
+  def ratioHead = "Ratio" ~> "(" ~> rep1sep(headAtom, ",") <~ ")" ^^ {
+    InferenceRuleHead(FactorFunction.Ratio, _)
+  }
```
Shouldn't we try to share `=>` syntax for these Linear and Ratio semantics?  Honestly, I don't have a good idea, but maybe with the help of annotation syntax?

```
@linear
P(x,y) => Q(x) :- ... .

@ratio
P(x,y) => Q(x) :- ... .
```

netj commented on src/main/scala/org/deepdive/ddlog/DeepDiveLogParser.scala:
```diff
@@ -265,43 +267,82 @@ class DeepDiveLogParser extends JavaTokenParsers {
+  def orHead    = headAtom ~ "v" ~ rep1sep(headAtom, "v") ^^ { case (a ~ _ ~ b) =>
+    InferenceRuleHead(FactorFunction.Or, a +: b)
+  }
+  def multinomialHead = "Multinomial" ~> "(" ~> rep1sep(headAtom, ",") <~ ")" ^^ {
```
These multinomial factors always makes me confusing when written like predicates.  Is there any better way to write them that still looks okay mixed in logical formulae?

feiranwang replied:
I didn't manage to come up with a nice symbol... so I used the function name here

netj commented on src/main/scala/org/deepdive/ddlog/DeepDiveLogParser.scala:
```diff
@@ -265,43 +267,82 @@ class DeepDiveLogParser extends JavaTokenParsers {
+  def ratioHead = "Ratio" ~> "(" ~> rep1sep(headAtom, ",") <~ ")" ^^ {
+    InferenceRuleHead(FactorFunction.Ratio, _)
+  }
+  def inferenceRuleHead = implyHead | equalHead | andHead | orHead | multinomialHead | linearHead | ratioHead | isTrueHead
```
I'm curious what are the pros when these inference rule heads are written separately.  I feel like keeping all inline here might be simpler.

netj commented on src/main/scala/org/deepdive/ddlog/DeepDiveLogParser.scala:
```diff
@@ -319,6 +360,8 @@ class DeepDiveLogParser extends JavaTokenParsers {
     }
   }
 
+  def error(meesage: String) = throw new RuntimeException(meesage)
```
`sys.error()`?

feiranwang replied:
Seems sys.error is same as throwing a RuntimeException?

netj replied:
Right.  So this function seems redundant.

netj commented:
Looks great overall!  Super excited to get this in.  Please see my inline comments.  You may ignore most of the function extraction suggestions on Compiler, which probably should be done in a separate PR.  However, let's nail down some of the syntax design issues.

feiranwang committed:
Bug fix and clean up

1. Fix UNION ALL bug in generating SQL queries.
2. Rewrite smoke and spouse example using new imply head
3. Code clean up

netj commented on src/main/scala/org/deepdive/ddlog/DeepDiveLogParser.scala:
```diff
@@ -265,43 +267,90 @@ class DeepDiveLogParser extends JavaTokenParsers {
+    }
+  | "@semantics(ratio)" ~> implyHeadAtoms ^^ {
+      InferenceRuleHead(FactorFunction.Ratio, _)
+    }
```
Can you move these imply cases up right below the first one with logical semantics?  I feel they should be grouped together to also make it clear why `implyHeadAtoms` is there.

netj commented on test/expected-output-test/spouse_example_new_inference/input.ddl:
```diff
@@ -55,29 +51,27 @@ function ext_has_spouse over like ext_has_spouse_input
-  has_spouse(rid2),
+@mode(inc)
+@weight(3.0)
+has_spouse(rid2) => has_spouse(rid) :-
```
Should we change this back to an Equals factor since we now support it?

netj commented:
Please see my comments (minor ones).  Let's merge this and open an issue for revisiting the `Multinomial` factor syntax.

feiranwang committed:
Update spouse example test with equal function

feiranwang commented:
Updated.

netj commented on src/main/scala/org/deepdive/ddlog/DeepDiveLogParser.scala:
```diff
@@ -253,8 +248,15 @@ class DeepDiveLogParser extends JavaTokenParsers {
+    case (name ~ args) => RuleAnnotation(name, args)
+  }
+
+  def ruleAnnotations = rep(ruleAnnotation)
```
I was checking which issues this PR was closing, and it seems the ruleAnnotations aren't really being parsed or attached to the parse tree yet, right?

feiranwang replied:
In 68581a3df432c4c9be5c263dc1dc697a1d834418 I tried the general annotation for extraction and inference rules, but it seems awkward, so in cf5759fab85cfd507d6abaf6b9c25ad71a0e56cd I switched back to the current style. These lines are for future use of rule annotations.

netj replied:
Okay, sounds good.  So #20 stays open.  When doing this later, let's export the annotations on the rules into JSON as well.  That means we probably need to come up with a nice and compact way to represent the parse tree for rules in JSON first.

netj commented:
Looks great! Merging..
